トランザクションロールバック時にランタイム例外が発生することでローカルトランザクションの状態が壊れる

LocalTransaction#rollbackInternal 内、 try - finally ブロック外での LocalTransactionContext#getConnection がランタイム例外をスローするとスレッドローカルに LocalTransactionContext インスタンスへの参照が残った状態になり、以降同一スレッドにおけるトランザクションが開始できなくなります。

トランザクションが開始された状態で JDBC 接続がサーバーから切断された場合等にこの事象が起きます。このときローカルトランザクションはロールバックしようとしますが、前述の通りロールバック処理自体が失敗し、 LocalTransaction オブジェクトのスレッドローカルに切断されたコネクションを参照した LocalTransactionContext が残ります。

```
org.seasar.doma.jdbc.JdbcException: [DOMA2056] トランザクション分離レベルの取得に失敗しました。原因は次のものです。java.sql.SQLException: Connection is closed
	at org.seasar.doma.jdbc.tx.LocalTransaction.lambda$beginInternal$0(LocalTransaction.java:212)
	at org.seasar.doma.jdbc.tx.LocalTransactionContext.getConnection(LocalTransactionContext.java:78)
	at org.seasar.doma.jdbc.tx.LocalTransaction.rollbackInternal(LocalTransaction.java:346)
	at org.seasar.doma.jdbc.tx.LocalTransaction.rollback(LocalTransaction.java:328)
	at org.seasar.doma.jdbc.tx.LocalTransactionManager.executeInTransaction(LocalTransactionManager.java:251)
	at org.seasar.doma.jdbc.tx.LocalTransactionManager.requiredInternal(LocalTransactionManager.java:96)
	at org.seasar.doma.jdbc.tx.LocalTransactionManager.required(LocalTransactionManager.java:75)
	...
Caused by: java.sql.SQLException: Connection is closed
	at com.zaxxer.hikari.pool.ProxyConnection$ClosedConnection.lambda$getClosedConnection$0(ProxyConnection.java:475)
	at com.sun.proxy.$Proxy96.getTransactionIsolation(Unknown Source)
	at com.zaxxer.hikari.pool.HikariProxyConnection.getTransactionIsolation(HikariProxyConnection.java)
	at org.seasar.doma.jdbc.tx.LocalTransaction.lambda$beginInternal$0(LocalTransaction.java:209)
	... 21 common frames omitted
```

--- a/src/main/java/org/seasar/doma/jdbc/tx/LocalTransaction.java
+++ b/src/main/java/org/seasar/doma/jdbc/tx/LocalTransaction.java
@@ -202,8 +202,8 @@ public class LocalTransaction {
             throw new TransactionAlreadyBegunException(id);
         }
         context = getLocalTransactionContext();
-        context.begin((connection) -> {
-            assertNotNull(connection);
+        context.begin(() -> {
+            Connection connection = JdbcUtil.getConnection(dataSource);
             int transactionIsolation;
             try {
                 transactionIsolation = connection.getTransactionIsolation();
@@ -241,9 +241,7 @@ public class LocalTransaction {
      * @return ローカルトランザクションコンテキスト
      */
     protected LocalTransactionContext getLocalTransactionContext() {
-        LocalTransactionContext context = new LocalTransactionContext(() -> {
-            return JdbcUtil.getConnection(dataSource);
-        });
+        LocalTransactionContext context = new LocalTransactionContext();
         localTxContextHolder.set(context);
         return context;
     }

--- a/src/main/java/org/seasar/doma/jdbc/tx/LocalTransactionContext.java
+++ b/src/main/java/org/seasar/doma/jdbc/tx/LocalTransactionContext.java
@@ -17,13 +17,11 @@ package org.seasar.doma.jdbc.tx;
 
 import static org.seasar.doma.internal.util.AssertionUtil.assertNotNull;
 
-import java.sql.Connection;
 import java.sql.Savepoint;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.function.Function;
 import java.util.function.Supplier;
 
 import org.seasar.doma.message.Message;
@@ -38,50 +36,37 @@ public class LocalTransactionContext {
 
     private final Map<String, Savepoint> savepointMap = new HashMap<String, Savepoint>();
 
-    private final Supplier<Connection> connectionSupplier;
-
-    private Connection connection;
-
     private LocalTransactionConnection localTxConnection;
 
-    private Function<Connection, LocalTransactionConnection> connectionInitializer;
+    private Supplier<LocalTransactionConnection> localTxConnectionSupplier;
 
     private String id;
 
     private boolean rollbackOnly;
 
-    LocalTransactionContext(Supplier<Connection> connectionSupplier) {
-        assertNotNull(connectionSupplier);
-        this.connectionSupplier = connectionSupplier;
-    }
-
-    void begin(
-            Function<Connection, LocalTransactionConnection> connectionInitializer) {
-        assertNotNull(connectionInitializer);
-        id = String.valueOf(System.identityHashCode(connectionInitializer));
-        this.connectionInitializer = connectionInitializer;
+    void begin(Supplier<LocalTransactionConnection> localTxConnectionSupplier) {
+        assertNotNull(localTxConnectionSupplier);
+        id = String.valueOf(System.identityHashCode(localTxConnectionSupplier));
+        this.localTxConnectionSupplier = localTxConnectionSupplier;
     }
 
     void end() {
         id = null;
-        connectionInitializer = null;
+        localTxConnectionSupplier = null;
     }
 
     LocalTransactionConnection getConnection() {
         if (localTxConnection == null) {
-            if (connection == null) {
-                connection = connectionSupplier.get();
-            }
-            if (connectionInitializer == null) {
+            if (localTxConnectionSupplier == null) {
                 throw new TransactionNotYetBegunException(Message.DOMA2048);
             }
-            localTxConnection = connectionInitializer.apply(connection);
+            localTxConnection = localTxConnectionSupplier.get();
         }
         return localTxConnection;
     }
 
     boolean hasConnection() {
-        return connection != null;
+        return localTxConnection != null;
     }
 
     Savepoint getSavepoint(String savepointName) {

--- a/src/test/java/org/seasar/doma/jdbc/tx/LocalTransactionContextTest.java
+++ b/src/test/java/org/seasar/doma/jdbc/tx/LocalTransactionContextTest.java
@@ -19,9 +19,9 @@ import java.sql.Connection;
 import java.sql.SQLException;
 import java.sql.Savepoint;
 
-import junit.framework.TestCase;
-
 import org.seasar.doma.internal.jdbc.mock.MockConnection;
+
+import junit.framework.TestCase;
 
 /**
  * @author taedium
@@ -32,8 +32,7 @@ public class LocalTransactionContextTest extends TestCase {
     public void testReleaseAndGetSavepoint() throws Exception {
         try (LocalTransactionConnection connection = new LocalTransactionConnection(
                 new MockConnection(), Connection.TRANSACTION_READ_COMMITTED)) {
-            LocalTransactionContext context = new LocalTransactionContext(
-                    () -> connection);
+            LocalTransactionContext context = new LocalTransactionContext();
             context.addSavepoint("1", new MySavepoint("1"));
             context.addSavepoint("2", new MySavepoint("2"));
             context.addSavepoint("3", new MySavepoint("3"));

